### PR TITLE
Adding API Key subscription validation feature

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
Provide a new feature to validate API Key Subscriptions.
Provide configuration to enable/disable this feature.

## WUM Related PR
https://github.com/wso2-support/carbon-apimgt/pull/3535

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/keys/APIKeyDataStore.java [hunk 1]
```diff
@@ -97,6 +97,18 @@ APIKeyValidationInfoDTO validateSubscription(String context, String version, Str
                                                  String tenantDomain, String keyManager)
             throws APISecurityException;
 
+    /**
+     * Validate API subscriptions.
+     *
+     * @param context Context of an API
+     * @param version A valid version of the API
+     * @param appId ApplicationId
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/keys/APIKeyDataStore.java:105]: Shall we add tenant domain too?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/GatewayUtils.java [hunk 2]
```diff
@@ -908,6 +935,18 @@ public static boolean isGatewayTokenCacheEnabled() {
         return true;
     }
 
+    public static boolean isAPIKeySubscriptionValidationEnabled() {
+        try {
+            APIManagerConfiguration config = ServiceReferenceHolder.getInstance().getAPIManagerConfiguration();
+            String subscriptionValidationEnabled = config.getFirstProperty(APIConstants.API_KEY_SUBSCRIPTION_VALIDATION_ENABLED);
+            return Boolean.parseBoolean(subscriptionValidationEnabled);
+        } catch (Exception e) {
+            log.error("Did not found valid API Key Subscription Validation Enabled configuration. " +
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/GatewayUtils.java:944]: Let's correct as "Did not find..."

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/KeyValidationHandler.java [hunk 3]
```diff
@@ -45,6 +45,16 @@ boolean validateSubscription(TokenValidationContext tokenValidationContext)
      * @param consumerKey
      */
     APIKeyValidationInfoDTO validateSubscription(String apiContext, String apiVersion, String consumerKey, String keyManager);
+
+    /**
+     * Validate Subscriptions by API context, API version and Application ID
+     *
+     * @param apiContext
+     * @param apiVersion
+     * @param appId
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/KeyValidationHandler.java:54]: Missing return statement in java docs

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyValidationService.java [hunk 4]
```diff
@@ -539,6 +539,24 @@ public boolean validateScopes(TokenValidationContext tokenValidationContext, Str
         return keyValidationHandler.validateScopes(tokenValidationContext);
     }
 
+    /**
+     * Validates the subscriptions of a particular API.
+     *
+     * @param context     Requested context
+     * @param version Version of the API
+     * @param appId Application ID
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyValidationService.java:547]: Let's add tenant domain and exception details to java docs

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/keys/APIKeyDataStore.java (hunk 1 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/keys/APIKeyDataStore.java:105]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/KeyValidationHandler.java (hunk 3 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/KeyValidationHandler.java:54]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyValidationService.java (hunk 4 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/service/APIKeyValidationService.java:547]: resolved, removed unused exceptions throws

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/GatewayUtils.java (hunk 2 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/utils/GatewayUtils.java]: resolved

**Review by npamudika — APPROVED** [review]

**Pull request merged**